### ISSUE TITLE
How to perform tracking independently

### ISSUE THREAD

**Issue by elbadaernU404** [ultralytics/ultralytics#23261]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I hope to execute the prediction and tracking tasks separately to enhance speed. By storing the prediction results and then passing them to the tracking module, these tasks can run independently. Is this feasible? Below is the code I used for testing. The results seem to fall short of expectations. I tried using both bytetrack and botsort, but the returned results were different。

**Comment by UltralyticsAssistant** [comment]
👋 Hello @elbadaernU404, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it (your snippet/results description is helpful, but we’ll need a complete MRE we can run end-to-end).

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

A few details that would help us understand your “predict first, track later” workflow and why ByteTrack vs BoT-SORT differs for you:
-How are you serializing/storing the prediction outputs (format, precision, coordinate space, confidences, class IDs)?
-What exact inputs are you passing into the tracking module afterward (per-frame detections, frame rate/time deltas, original image shapes)?
-Please include the exact commands or Python calls used, plus a small sample of the stored predictions (one or two frames) and the resulting track outputs

This is an automated response 🤖✨ — an Ultralytics engineer will also assist soon.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by elbadaernU404** [comment]
Here is my test code. How can I modify it?
```python
import cv2
import torch
from ultralytics import YOLO
from ultralytics.trackers.byte_tracker import BYTETracker
from ultralytics.trackers.bot_sort import BOTSORT
from ultralytics.trackers.track import on_predict_start, on_predict_postprocess_end
from ultralytics.utils import YAML, IterableSimpleNamespace
from ultralytics.utils.checks import check_yaml
from ultralytics.utils.ops import xywh2xyxy
from ultralytics.utils.plotting import save_one_box

# Load the model and run the tracker with a custom configuration file
model = YOLO("yolo26n.pt")
tracker_yaml = check_yaml('./ultralytics/cfg/trackers/customtrack.yaml')
cfg = IterableSimpleNamespace(**YAML.load(tracker_yaml))
tracker = BOTSORT(args=cfg, frame_rate=30)
im = cv2.imread('./input/frames/1.png')

results = model.predict(
    source='./input/frames/1.png',
    tracker='./ultralytics/cfg/trackers/customtrack.yaml',
    classes=[0],
    conf=0.6,
    stream=False,
    stream_buffer=False,
    vid_stride=1,
    batch=1,
    device='cuda:1',
    half=True,
    # save=True,
    # project='output',
    # name='sort'
    # embed=['C5']
)

for result in results:
    tracked_objects = tracker.update(result.boxes.cpu().numpy(), im, getattr(result, "feats", None))
    idx = tracked_objects[:, -1].astype(int)
    box = tracked_objects[:, :-1]
    print(idx, box)
```

**Comment by raimbekovm** [comment]
Output format of `tracker.update()` is `[x1, y1, x2, y2, track_id, conf, cls, det_idx]`.

Your parsing:
```python
idx = tracked_objects[:, -1]  # this is det_idx, not track_id
box = tracked_objects[:, :-1]  # 7 columns instead of 4
```

Correct:
```python
track_ids = tracked_objects[:, 4]
boxes = tracked_objects[:, :4]
```

**Comment by elbadaernU404** [comment]
> Output format of `tracker.update()` is `[x1, y1, x2, y2, track_id, conf, cls, det_idx]`.
> 
> Your parsing:
> 
> idx = tracked_objects[:, -1]  # this is det_idx, not track_id
> box = tracked_objects[:, :-1]  # 7 columns instead of 4
> Correct:
> 
> track_ids = tracked_objects[:, 4]
> boxes = tracked_objects[:, :4]

Thank you for your reply! I found that tracking is possible, but how can I obtain the features of reid directly obtained through tracker.update() operation and use the gmc function? Also, I noticed that persist=True does not take effect when directly using tracker.update()

**Comment by elbadaernU404** [comment]
> > 输出格式`tracker.update()`为`[x1, y1, x2, y2, track_id, conf, cls, det_idx]`。
> > 您的解析：
> > idx = tracked_objects[:, -1] # 这是 det_idx，不是 track_id 
> > box = tracked_objects[:, :-1] # 7 列而不是 4 列
> > 正确：
> > track_ids = tracked_objects[:, 4] 
> > boxes = tracked_objects[:, :4]
> 
> 感谢您的回复！我发现跟踪是可行的，但是如何才能直接获取通过 tracker.update() 操作更新的 reid 特征并使用 gmc 函数呢？另外，我注意到直接使用 tracker.update() 时 persist=True 似乎不起作用。

I modified the project source code to implement this function：
```python
    def init_track(self, results, img: np.ndarray | None = None) -> list[BOTrack] | tuple:
        """Initialize object tracks using detection bounding boxes, scores, class labels, and optional ReID features."""
        if len(results) == 0:
            return []
        bboxes = results.xywhr if hasattr(results, "xywhr") else results.xywh
        bboxes = np.concatenate([bboxes, np.arange(len(bboxes)).reshape(-1, 1)], axis=-1)
        if self.args.with_reid and self.encoder is not None:
            features_keep = self.encoder(img, bboxes)
            return features_keep, [BOTrack(xywh, s, c, f) for (xywh, s, c, f) in zip(bboxes, results.conf, results.cls, features_keep)]
        else:
            return [BOTrack(xywh, s, c) for (xywh, s, c) in zip(bboxes, results.conf, results.cls)]
```
Obtain features in this step and output to `tracker.update()`:
```python
    features, tracked_objects = tracker.update(result.boxes.cpu().numpy(), im, getattr(result, "feats", None))
```

**Comment by elbadaernU404** [comment]
> > > 这是输出格式`tracker.update()`为`[x1, y1, x2, y2, track_id, conf, cls, det_idx]`。
> > > 您的解析：
> > > idx = tracked_objects[:, -1] # det_idx，不是 track_id 
> > > box = tracked_objects[:, :-1] # 7 列不是 4 列
> > > 正确：
> > > track_ids = tracked_objects[:, 4] 
> > > box = tracked_objects[:, :4]
> > 
> > 
> > 谢谢您的回复！我发现跟踪是可以安装的，但是如何才能直接获取通过 tracker.update() 操作更新的 reid 特征并使用 gmc 函数呢？ 另外，我注意到直接使用 tracker.update() 时 persist=True 似乎不行。
> 
> 我修改了项目源代码以实现此功能：
> 
>     def init_track(self, results, img: np.ndarray | None = None) -> list[BOTrack] | tuple:
>         """Initialize object tracks using detection bounding boxes, scores, class labels, and optional ReID features."""
>         if len(results) == 0:
>             return []
>         bboxes = results.xywhr if hasattr(results, "xywhr") else results.xywh
>         bboxes = np.concatenate([bboxes, np.arange(len(bboxes)).reshape(-1, 1)], axis=-1)
>         if self.args.with_reid and self.encoder is not None:
>             features_keep = self.encoder(img, bboxes)
>             return features_keep, [BOTrack(xywh, s, c, f) for (xywh, s, c, f) in zip(bboxes, results.conf, results.cls, features_keep)]
>         else:
>             return [BOTrack(xywh, s, c) for (xywh, s, c) in zip(bboxes, results.conf, results.cls)]
> 在此步骤中获取特征并输出到`tracker.update()`：
> 
>     features, tracked_objects = tracker.update(result.boxes.cpu().numpy(), im, getattr(result, "feats", None))

But I still don't quite understand how to start gmc, and how to use persist=True

**Comment by raimbekovm** [comment]
GMC should already work - just pass `img` to `tracker.update(results, img=frame)`.

For ReID, `model='auto'` won't work with direct `tracker.update()` - it expects features from predictor hooks. Use external model instead:
```python
args.model = 'yolo11n-cls.pt'
```
Then access features via `track.smooth_feat` after update.

As for persist - tracker keeps state between `update()` calls by default. Call `tracker.reset()` when switching to a new video.

**Comment by glenn-jocher** [comment]
`persist=True` only applies to the high-level `model.track()` API; if you’re calling `tracker.update()` directly, “persistence” is simply reusing the same `tracker` instance across frames and only calling `tracker.reset()` when you switch to a new video (this is what the built-in tracking hooks do in [`on_predict_postprocess_end()`](https://docs.ultralytics.com/reference/trackers/track/)). GMC doesn’t need a separate “start” call either—`BOTSORT` runs it automatically whenever you pass the current frame as `img` into `update()` (see the `img`/GMC path in [`BYTETracker.update()`](https://docs.ultralytics.com/reference/trackers/byte_tracker/) and the BoT-SORT implementation in [`bot_sort.py`](https://docs.ultralytics.com/reference/trackers/bot_sort/)).

ReID embeddings are not included in the returned `tracks` array, so don’t patch `init_track()`; after `update()` you can read them from `tracker.tracked_stracks` (each `BOTrack` keeps `curr_feat`/`smooth_feat`), and if you’re running tracking “independently” you generally should set a ReID `model` in your tracker YAML (e.g. `yolo26n-cls.pt`) instead of `model: auto`, since `auto` expects predictor-extracted features (see the [tracking mode docs](https://docs.ultralytics.com/modes/track/)).

```python
tracks = tracker.update(det, img=frame, feats=feats)  # feats can be None if using a *-cls.pt ReID model
for t in tracker.tracked_stracks:
    if t.is_activated:
        track_id = t.track_id
        reid_feat = t.smooth_feat  # or t.curr_feat
```

**Comment by elbadaernU404** [comment]
> `persist=True` only applies to the high-level `model.track()` API; if you’re calling `tracker.update()` directly, “persistence” is simply reusing the same `tracker` instance across frames and only calling `tracker.reset()` when you switch to a new video (this is what the built-in tracking hooks do in [`on_predict_postprocess_end()`](https://docs.ultralytics.com/reference/trackers/track/)). GMC doesn’t need a separate “start” call either—`BOTSORT` runs it automatically whenever you pass the current frame as `img` into `update()` (see the `img`/GMC path in [`BYTETracker.update()`](https://docs.ultralytics.com/reference/trackers/byte_tracker/) and the BoT-SORT implementation in [`bot_sort.py`](https://docs.ultralytics.com/reference/trackers/bot_sort/)).
> 
> ReID embeddings are not included in the returned `tracks` array, so don’t patch `init_track()`; after `update()` you can read them from `tracker.tracked_stracks` (each `BOTrack` keeps `curr_feat`/`smooth_feat`), and if you’re running tracking “independently” you generally should set a ReID `model` in your tracker YAML (e.g. `yolo26n-cls.pt`) instead of `model: auto`, since `auto` expects predictor-extracted features (see the [tracking mode docs](https://docs.ultralytics.com/modes/track/)).
> 
> tracks = tracker.update(det, img=frame, feats=feats)  # feats can be None if using a *-cls.pt ReID model
> for t in tracker.tracked_stracks:
>     if t.is_activated:
>         track_id = t.track_id
>         reid_feat = t.smooth_feat  # or t.curr_feat

Thank you for your detailed explanation! It has been of great help to me in reading the source code. I have already noticed that sometimes the ReID embeddings returned directly from the `tracker.update()` are empty arrays after I patch `init_track()`.

Could you further explain the difference between `smooth_feat` and `curr_feat`? I understand that `curr_feat` represents the target feature of the current frame, but what does `smooth_feat` stand for? They seem almost identical.Thank you very much:D

**Comment by raimbekovm** [comment]
`smooth_feat` is an EMA of features over time (`alpha=0.9`), `curr_feat` is just the current frame.

In matching, tracks use `smooth_feat` (stable identity) while detections use `curr_feat` (single observation). Smoothing helps handle occlusion and noise.

**Comment by elbadaernU404** [comment]
> `smooth_feat`是特征随时间变化的 EMA（`alpha=0.9`），`curr_feat`只是当前帧。
> 
> 在匹配过程中，轨迹使用`smooth_feat`（稳定身份），而检测使用`curr_feat`（单次观测）。平滑处理有助于处理遮挡和噪声。

Thank you for your guidance.I understand now, Have a nice day:D

**Comment by glenn-jocher** [comment]
Glad that cleared it up; if you ever need the tracker to react faster/slower to appearance changes you can adjust the feature smoothing via `BOTrack.alpha` (and the history length via `feat_history`) as shown in the [BoT-SORT tracker reference](https://docs.ultralytics.com/reference/trackers/bot_sort/), and if any new tracking/ReID questions come up please open a new issue for visibility since this one is closed.

**Comment by elbadaernU404** [comment]
> Glad that cleared it up; if you ever need the tracker to react faster/slower to appearance changes you can adjust the feature smoothing via `BOTrack.alpha` (and the history length via `feat_history`) as shown in the [BoT-SORT tracker reference](https://docs.ultralytics.com/reference/trackers/bot_sort/), and if any new tracking/ReID questions come up please open a new issue for visibility since this one is closed.

Okey.Thank you very much for your help

**Comment by glenn-jocher** [comment]
You’re welcome—happy to help, and if anything else comes up with tracking/ReID just open a new issue so the team sees it.
